File: python/Chapter 2/Question2_4.py
from classes.LinkedList import *

def partition(linkedlist, x):
    if linkedlist.head != None:
        p1 = linkedlist.head
        p2 = linkedlist.head.next
        while p2 != None:
            if p2.value < x:
                p1.next = p2.next
                p2.next = linkedlist.head
                linkedlist.head = p2
                p2 = p1.next
            else:
                p1 = p1.next
                p2 = p1.next



#----------------test-----------------
L = randomLinkedList(10, 0, 50)
x = 25

print L, " , x=25"   
partition(L, x)
print L 

File: c++/Chapter 2/Question2_4.cpp
#include <iostream>
#include "LinkedListClass/LinkedList.h"
#include "../Chapter 1/Question2_4.h"

// runs in O(n) time with O(n) space
// there are various ways to do this, but I end up deleting the original list
// and returning the merged lower and upper value lists
Node* partition(Node* head, int val)
{
    if (head == NULL)
    {
        return nullptr;
    }

    LinkedList* upper = new LinkedList();
    LinkedList* lower = new LinkedList();
    Node* current = head;

    while (current)
    {
        if (current->data < val)
        {
            lower->insert(current->data);
        }
        else
        {
            upper->insert(current->data);
        }

        current = current->next;
    }

    // reset current back to the head
    current = lower->head;
    Node* lowNode = lower->head;

    // find the end of the lower list
    while (lowNode->next)
    {
        lowNode = lowNode->next;
    }
    lowNode->next = upper->head;

    return current;
}

int Question2_4::run()
{
    LinkedList* myList = new LinkedList();
    myList->insert(5);
    myList->insert(14);
    myList->insert(17);
    myList->insert(22);
    myList->insert(16);
    myList->insert(18);
    myList->insert(2);
    myList->insert(1);
    myList->insert(5);

    std::cout << "The original list is: ";
    myList->display();

    Node* newHead = partition(myList->head, 13);
    delete myList;

    LinkedList* newList = new LinkedList();
    newList->head = newHead;

    std::cout << "The list partitioned around value 13 is: ";
    newList->display();
    delete newList;
  
    return 0;
}